mesh/mesh_connections.hpp

```diff
@@ -0,0 +1,256 @@
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+class Mesh;
+
+/** @brief Struct representing the ID number of a geometrical mesh entity.  In 
+    MFEM meshes there is currently support for 5 entity kinds that all have their
+    own id numberings:  vertices, edges, faces, elements, and boundary elements. 
+    These kinds are represented by the @a dim and @a boundary members.  For
+    instance a mesh edge on the boundary will have dim=1 and boundary=true.  
+    This struct is used to describe the ID numbers recieved and returned by the 
+    \ref MeshConnections class.**/
+struct MeshEntityID
+{
+   int dim;
+   bool boundary;
+   int id;
+}
+
+/** @brief Struct representing plura ID numbers of geometrical mesh entities of the 
+    same kind.  In MFEM meshes there is currently support for 5 (4 in 2D entity kinds 
+    that all have their own id numberings:  vertices, edges, faces, elements, and boundary 
+    elements.   These types are represented by the @a dim and @a boundary members.  For
+    instance faces in a 3D mesh interior will have dim=2 and boundary=false.**/
+struct MeshEntityIDs
+{
+   int dim;
+   bool boundary;
+   Array<int> ids;      
+}
+
+/** @brief Class that exposes connections between Mesh entities such as
+    vertices, edges, faces, and volumes.  This is done without assuming 
+    the elements have a particular dimension or type and will support 
+    dimensions > 3 in the future.  Entity ID numbers are described in the 
+    \ref MeshEntityID and \ref MeshEntityIDs sturcts.  The connections are defined 
+    on the Mesh object given at the time of construction, and Table objects are built 
+    lazily as they are required so the first query of a paticular kind may be a lot slower
+    than the following queries.  It is worth noting that the our naming convention of child
+    and parent connections actually describes ancestor/decendent relationship eg. a node can 
+    e a "child" of edges, faces, and volumes.**/
+class MeshConnections
+{
+   private:
+   /// The mesh object that the connections are defined on
+   Mesh *mesh;
+
+   /// 2D array of tables describing connections between non-boundary entities
+   Array<Array<Table*>> T;
+
+   /// 2D array of tables describing connections between boundary elements and other entities
+   Array<Array<Table*>> BT;
+
+   public:
+   /** @brief Sets the Mesh object connections are defined on. */
+   /** @note When this method is called, no connection tables
+       are generated so they will all be null.  They will be generated
+       lazily as access is needed.**/
+   MeshConnections(Mesh *m);
+   
```

Comment:
I'm not sure what you mean about the boundary attribute being  not straightforward to provide.  To index into an arbitrary entity you need the dimension of the entity, whether it is boundary entity, and the index number.  So you can index it with an EntityIndex struct which has all 3 of those.  Anything less than than and you don't always have enough info to refer to the correct entity.  I think I am missing what you are saying, can you elaborate?

Reply:
I think I am getting confused in the different types of indices that we have. In `Mesh::FaceInformation` the `index` is independent of boundary or not, it is the "global face index". If the `MeshConnections` faces use different indices from the ones in `Mesh::FaceInformation`, that could make things quite confusing for users.

The problem I think is that there is different kinds of indices associated to the same mesh entity, e.g boundary faces have either "boundary face index" and "global face index". Other functions have index type requirements implicitly suggested by the function name, e.g. `Mesh::GetBdrAttribute(int i)` expects a "boundary face index" whereas `Mesh::GetAttribute(int i)` expects a global index. According to the methods used, users may either need the "boundary" index or the "global" index of boundary faces, e.g treating boundary conditions, or treating DG methods.

What I would suggest to consider in a future PR is to type the "boundary face indices":
```c++
struct BoundaryFaceIndex
{
   int index.

   explicit BoundaryFaceIndex(int index): index(index) { }
   
   int operator int()
   {
      return index;
   }
};
```
Then we can apply the change everywhere, for instance the interface of
```c++
Mesh::GetBdrAttribute(int i)
```
to
```c++
Mesh::GetBdrAttribute(const BoundaryFaceIndex &i)
```
and deprecate the previous version.
I would also expose methods that maps both ways "boundary face index" and "global face index", for instance
```c++
std::optional<BoundaryFaceIndex> Mesh::GetBoundaryFaceIndex(int global_face_index) const
```
and
```c++
int  Mesh::GetGlobalFaceIndex(BoundaryFaceIndex bdr_face_index) const
```
.

Therefore, I am wondering if we wouldn't want to always return the "global face index" (even for boundary faces) in all the `MeshConnections` methods, to always return indices of the same "kind", and let users be responsible for calling `std::optional<BoundaryFaceIndex> Mesh::GetBoundaryFaceIndex(int global_face_index)`.

Reply:
Yeah I haven't changed anything about how anything is numbered in the Mesh class here.  I just rounded up all the things that were using the tables and put them in one place with a consistent interface.  After we get this in there I think we should take a long hard look at all of the indices we have floating around in there and unifying them.

Reply:
My point was that even if the tables are working with both "global face indices" and "boundary face indices", I think we should unify the interface and only use "global face indices", and treat the "boundary face index" as an implementation detail. I think it is too bug-prone to mix "global face indices" for non-boundary faces with "boundary face indices" for boundary faces. But it also depends on users' usage, I don't know if it is more natural to manipulate "boundary face indices" or "global face indices" for boundary faces, my intuition is that it depends on the situation, therefore a unified interface only on "global face indices" makes more sense to me (with the proposed utility functions to go between "boundary face indices" and "global face indices). Once we only use "global face indices" then the `boundary` attribute becomes obsolete in many cases I think.

Reply:
We need to be able to index into the array of boundary elements.  We could special case them, but I don't think that makes the interface nicer.  We still want to be able to do queries between boundary elements and non boundary elements, so you will still need the boundary flag in both the in and out parameters of the so it would just move from the EntityIndex structs out to the method parameters.  So really it's just a matter of which numbering those queries return, the numbers that the user is more likely to need are the boundary element numbers, but I do agree that it hangs together nicely if there is only one index space for each dimensional object.  Either way I think we need mehtods to go back and forth between boundary element and standard face numberings.

Reply:
> We need to be able to index into the array of boundary elements.

This is the purpose of:
```c++
std::optional<BoundaryFaceIndex> Mesh::GetBoundaryFaceIndex(int global_face_index) const

int  Mesh::GetGlobalFaceIndex(BoundaryFaceIndex bdr_face_index) const
```

>We could special case them, but I don't think that makes the interface nicer.

I think it does, because instead of mixing "boundary face indices" and "global face indices" in both the inputs and outputs, you only have "global face indices" without the need for `boundary` in `EntityID`, I think it makes the interface significantly simpler.

>We still want to be able to do queries between boundary elements and non boundary elements, so you will still need the boundary flag in both the in and out parameters of the so it would just move from the EntityIndex structs out to the method parameters.

I think there is a limitation with this, for instance when we want "either boundary or non-boundary". Also, having `boundary` as input parameter makes the interface more explicit IMHO, c.f. my remarks on the ambiguous semantic of `boundary` in some of the methods. Unless `boundary == false` means "either boundary or non-boundary".

>So really it's just a matter of which numbering those queries return, the numbers that the user is more likely to need are the boundary element numbers,

This is not so obvious to me, for instance, I had never used the "boundary face index" before, the boundaries in the DG are all handled through the "global face index". Also, users don't get access to the same functionality of MFEM with a "boundary face index" and "global face index". And finally, it is not only about the fact that the returned indices are of different kinds, but the input indices too. For instance, users that would want to iterate over the faces and look at the connectivity to apply a function `foo` would have to write code like this:
```c++
for ( int f = 0; f < fespace.GetNF(); f++ )
{
   auto face = mesh.GetFaceInformation(f);
   EntityID face_id;
   face_id.dim = mesh.Dimension() - 1;
   if ( face_id.boundary = face.IsBoundary() )
   {
      // Get the boundary face index somehow
      boundary_index = ??;
      face_id.index = boundary_index; // Warning: Using `f` would be a bug.
   }
   else
   {
      face_id.index = f;
   }
   MeshConnections mesh_conn(mesh);
   // Some function on the face
   foo( mesh_conn, face_id );
}
```
compared to
```c++
for ( int f = 0; f < fespace.GetNF(); f++ )
{
   EntityID face_id = { mesh.Dimension() - 1, f }; // boundary is unnecessary since it can be accessed through `mesh.GetFaceInformation(face_id.index).IsBoundary()`
   MeshConnections mesh_conn(mesh);
   // Some function on the face
   foo( mesh_conn, face_id );
}
```
Similar post-processing would be necessary for the output, for instance:
```c++
const Element* elem;
if ( entity_id.boundary )
{
   elem = mesh.GetBdrElement( entity_id.index );
}
else
{
   elem = mesh.GetElement( entity_id.index );
}
bar( elem );
```
compared to
```c++
bar( mesh.GetElement( entity_id.index ) );
```

>but I do agree that it hangs together nicely if there is only one index space for each dimensional object. Either way I think we need mehtods to go back and forth between boundary element and standard face numberings.

My understanding is that the non-boundary table contains the boundary elements, therefore I think the boundary table could just be used for implementing optimizations, or be exclusive to accesses with "boundary face indices".

Reply:
The code example you wrote up has some significant misunderstandings of how the numberings work in MFEM meshes and in my interface.  The boundary faces have both numberings floating around and given my current interface you can get at both numberings using the boundary flag.  So currently in the case where you want to run through all of the faces given the global face numbering we have:
```
for ( int f = 0; f < fespace.GetNF(); f++ )
{
   MeshConnections mesh_conn(mesh);
   EntityID face_id = { mesh.Dimension() - 1, false, f }; 
   // Some function on the face
   foo( mesh_conn, face_id );
}
```

If you want to run through all of the boundary faces in the boundary face numbering you would write:
```
for ( int f = 0; f < nbBoundaryFaces; f++ )
{
   MeshConnections mesh_conn(mesh);
   EntityID boundary_face_id = { mesh.Dimension() - 1, true, f }; 
   // Some function on the face
   foo( mesh_conn, boundary_face_id );
}
```

If you want to run through all of the boundary faces in the global face numbering you can write:
```
for ( int f = 0; f < nbBoundaryFaces; f++ )
{
   MeshConnections mesh_conn(mesh);
   EntityID face_id = { mesh.Dimension() - 1, false, mesh_conn.BoundaryElemToFaceIndex(f)}; 
   // Some function on the face
   foo(mesh_conn, face_id );
}
```
Or you can run through all of the faces and pick out the boundary faces in the global numbering like so:
```
for ( int f = 0; f < fespace.GetNF(); f++ )
{
   auto face = mesh.GetFaceInformation(f);
   if (face.IsBoundary() )
   {
      MeshConnections mesh_conn(mesh);
      EntityID face_id = {mesh.Dimension() - 1, false, f}
     
      // Some function on the face
      foo( mesh_conn, face_id );
  }
}
```
Perhaps the source if the confusion is the name of the flag boundary where boundary really is referring to the elements defined on those faces?  dim = mesh_dim-1, boundary = false indeed refers to all of the faces on the mesh

Reply:
You're right, I indeed didn't understand a few things. Would it be correct to say that what the `boundary` flag says is which "kind" of index the `index` is, i.e "global index" or "boundary index"? In that case wouldn't it be more explicit/self-documented to use an `enum` type? For instance:
```c++
/** "Boundary" indices are indexing the subset of boundary elements in a mesh, and are
      typically used with functions having "Bdr" in their name, e.g. Mesh::GetBdrElement(int i).
      "Global" indices are indexing the integrity of elements of a given dimension in a mesh,
      and are typically used with functions not containing "Bdr" in their name, e.g.
      Mesh::GetElement(int i). */
enum class IndexType : bool { Boundary, Global };

struct MeshEntityID
{
   int dim;
   IndexType index_type;
   int index;
}
```